A31-xds-routeaction-support.md

```diff
@@ -0,0 +1,184 @@
+can happen when clusters are no longer referenced by anything.  E.g. if the
+Config Selector has a reference to the deleted cluster, then when there are no
+more references to the config selector, the final reference to the cluster will
+be removed, triggering the Service Config update.
+
+Note that when a cluster is deleted in xDS, if the cds balancer were to queue
+RPCs in this scenario, wait-for-ready RPCs would be stuck indefinitely
+referencing the cluster, and it could never be removed.  Deleted clusters need
+to drop the RPC and return an UNAVAILABLE status.
+
+When new clusters are added to xDS, they will be added to both the Config
+Selector and Service Config simultaneously.  Here, precautions must be taken to
+ensure it is not possible for the Config Selector to be used before the Load
+Balancing Policy's Picker has been updated to be aware of the new cluster.  The
+channel must update the Load Balancing Policy before applying the Config
+Selector, and the Load Balancing Policy must provide a new Picker before its
+Update method returns back to the channel. If the Picker does not recognize a
+cluster name provided to it, which should not be possible when following the
+steps above in that order, it will drop the RPC with an INTERNAL status code.
+
+The following drawing shows which parts of the system will contain which current
+Load Balancing Policies.  The blue box represents the Config Selector and the
+red box represents the top-level Load Balancing Policy and its children.
+
+![xDS Architecture](A31_graphics/xds_architecture.png "xDS Architecture")
+
+Initially, only timeout-related `RouteAction` settings will be supported:
+<code>[timeout](http://google3/third_party/envoy/src/api/envoy/api/v2/route/route_components.proto?l=886&rcl=309630292)</code>
+and
+<code>[max_grpc_timeout](http://google3/third_party/envoy/src/api/envoy/api/v2/route/route_components.proto?l=979&rcl=309630292)</code>. <code>[grpc_timeout_offset](http://google3/third_party/envoy/src/api/envoy/api/v2/route/route_components.proto?l=988&rcl=309630292)</code>
+is used for shortening timeouts set by clients to ensure the timeout is reached
+before the client cancels the RPC, so supporting it in the client would be
+counter-productive.  If <code>max_grpc_timeout</code> is not present, the
+<code>timeout</code> field will specify the maximum RPC timeout for this route,
+with a default of 15s.  If <code>max_grpc_timeout</code> is present, then
+timeout is ignored and <code>max_grpc_timeout</code> limits the maximum timeout
+for RPCs on this route.  A value of 0 indicates no limit should be applied.  In
+all cases, the RPC timeout set by the application may never be exceeded due to
+these settings.
+
+## Rationale
+
+### Alternatives Considered
+
+Many alternatives were considered as part of this design.  Below are several of them:
+
+1.  Support xDS matchers in gRPC directly.
+    1.  This could lead to performance problems for existing users and would
```

Comment:
These sub-items can probably be bullets rather than numbered items.

Reply:
Done.